pkg/controller/direct/secretmanager/secret_controller.go

```diff
@@ -0,0 +1,381 @@
+				u.GetNamespace(), u.GetName(), id.Secret, resourceID)
+		}
+	}
+
+	if err = normalizeExternal(ctx, reader, u, obj); err != nil {
+		return nil, err
+	}
+
+	// Get secretmanager GCP client
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &Adapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+func (m *model) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id        *SecretManagerSecretIdentity
+	gcpClient *gcp.Client
+	desired   *krm.SecretManagerSecret
+	actual    *secretmanagerpb.Secret
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+func normalizeExternal(ctx context.Context, reader client.Reader, src client.Object, secret *krm.SecretManagerSecret) error {
+	if secret.Spec.Replication != nil {
+		if secret.Spec.Replication.LegacyAutomatic != nil {
+			if secret.Spec.Replication.LegacyAutomatic.CustomerManagedEncryption != nil {
+				kmsKeyRef := secret.Spec.Replication.LegacyAutomatic.CustomerManagedEncryption.KmsKeyRef
+
+				kmsKeyRef, err := refs.ResolveKMSCryptoKeyRef(ctx, reader, src, kmsKeyRef)
+				if err != nil {
+					return err
+				}
+				secret.Spec.Replication.LegacyAutomatic.CustomerManagedEncryption.KmsKeyRef = kmsKeyRef
+			}
+		}
+	}
+	if len(secret.Spec.TopicRefs) != 0 {
+		for _, topicRef := range secret.Spec.TopicRefs {
+			if topicRef.PubSubTopicRef != nil {
+				pubsubRef, err := refs.ResolvePubSubTopic(ctx, reader, src, topicRef.PubSubTopicRef)
+				if err != nil {
+					return err
+				}
+				topicRef.PubSubTopicRef.External = pubsubRef.String()
+			}
+		}
+	}
+	return nil
+
+}
+
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting SecretManagerSecret", "name", a.id.FullyQualifiedName())
+
+	req := &secretmanagerpb.GetSecretRequest{Name: a.id.FullyQualifiedName()}
+	secretpb, err := a.gcpClient.GetSecret(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting SecretManagerSecret %q: %w", a.id.FullyQualifiedName(), err)
+	}
+
+	a.actual = secretpb
+	return true, nil
+}
+
+func MergeMap(a, b map[string]string) map[string]string {
+	for k, v := range b {
+		a[k] = v
+	}
+	return a
+}
+
+func RemoveByPrefixes(a map[string]string, prefixes ...string) map[string]string {
+	for k, _ := range a {
+		for i := range prefixes {
+			if strings.HasPrefix(k, prefixes[i]) {
+				delete(a, k)
+			}
+		}
+	}
+	return a
+}
+
+func UserAnnotations(secret *krm.SecretManagerSecret) map[string]string {
+	annotations := MergeMap(secret.GetAnnotations(), secret.Spec.Annotations)
+	annotations = RemoveByPrefixes(annotations, "cnrm.cloud.google.com", "alpha.cnrm.cloud.google.com")
+	return annotations
+}
+
+func UserLabels(secret *krm.SecretManagerSecret) map[string]string {
```

Comment:
Nit: maybe "ComputeGCPLabels" / "ComputeGCPAnnotations" ?  Also I think we should immediately put these into a shared package, because they aren't something we want each service to write, we need to be compatible with existing behaviours.